File: runner/swift/Sources/Runner/APIRunner.swift
// Created by Yanjunhui

import Foundation
import MonoLiteSwift

/// API 测试运行器，通过直接调用 MonoLiteSwift API 执行测试
/// EN: API test runner that executes tests by directly calling MonoLiteSwift API
actor APIRunner {
    /// 数据库实例
    /// EN: Database instance
    private var db: Database?
    /// 数据库文件路径
    /// EN: Database file path
    private let dbPath: String

    /// 初始化运行器
    /// EN: Initialize the runner
    /// - Parameter dbPath: 数据库文件路径 / Database file path
    init(dbPath: String) {
        self.dbPath = dbPath
    }

    /// 打开数据库连接
    /// EN: Open database connection
    func open() async throws {
        db = try await Database(path: dbPath)
    }

    /// 关闭数据库连接
    /// EN: Close database connection
    func close() async throws {
        if let db = db {
            try await db.close()
            self.db = nil
        }
    }

    /// 运行单个测试用例
    /// EN: Run a single test case
    /// - Parameter tc: 测试用例 / Test case
    /// - Returns: 测试结果 / Test result
    func runTest(_ tc: TestCase) async -> TestResult {
        let start = Date()
        var result = TestResult(
            test_name: tc.name,
            language: "swift",
            mode: "api",
            success: false,
            duration_ms: 0
        )

        do {
            // 执行前置步骤 // EN: Execute setup steps
            try await executeSetup(tc)

            // 执行测试动作 // EN: Execute test action
            try await executeAction(tc, result: &result)

            result.success = true
        } catch {
            result.error = "\(error)"

            // 检查是否是预期的错误 // EN: Check if it's an expected error
            if let expectedError = tc.expected.error, result.error?.contains(expectedError) == true {
                result.success = true
            }
        }

        result.duration_ms = Int64(Date().timeIntervalSince(start) * 1000)
        return result
    }

    // MARK: - 前置步骤执行 / Setup Step Execution

    /// 执行测试用例的前置步骤
    /// EN: Execute setup steps for a test case
    /// - Parameter tc: 测试用例 / Test case
    private func executeSetup(_ tc: TestCase) async throws {
        guard let setup = tc.setup, !setup.isEmpty else { return }
        guard let db = db else { throw RunnerError.databaseNotOpened }

        let col = try await db.collection(tc.collection)

        for step in setup {
            let data = convertToBSONDocument(step.data.value)

            switch step.operation {
            case "insert":
                // 插入文档 // EN: Insert document
                _ = try await col.insertOne(data)
            case "createIndex":
                // 创建索引 // EN: Create index
                if let keys = (step.data.value as? [String: Any])?["keys"] as? [String: Any] {
                    let keysDoc = convertToBSONDocument(keys)
                    let optsDoc = BSONDocument()
                    _ = try await col.createIndex(keys: keysDoc, options: optsDoc)
                }
            default:
                break
            }
        }
    }

    // MARK: - 测试动作执行 / Test Action Execution

    /// 执行测试动作
    /// EN: Execute test action
    /// - Parameters:
    ///   - tc: 测试用例 / Test case
    ///   - result: 测试结果（引用传递） / Test result (passed by reference)
    private func executeAction(_ tc: TestCase, result: inout TestResult) async throws {
        guard let db = db else { throw RunnerError.databaseNotOpened }

        let col = try await db.collection(tc.collection)
        let action = tc.action

        let filter = action.filter.map { convertToBSONDocument($0.value) } ?? BSONDocument()
        let update = action.update.map { convertToBSONDocument($0.value) } ?? BSONDocument()
        let doc = action.doc.map { convertToBSONDocument($0.value) } ?? BSONDocument()
        let options = action.options?.value as? [String: Any] ?? [:]

        switch action.method {
        case "insertOne":
            // 插入单个文档 // EN: Insert single document
            _ = try await col.insertOne(doc)
            result.count = 1

        case "insertMany":
            // 批量插入文档 // EN: Batch insert documents
            if let docsArray = action.docs {
                let bsonDocs = docsArray.map { convertToBSONDocument($0.value) }
                let ids = try await col.insert(bsonDocs)
                result.count = Int64(ids.count)
            }

        case "find":
            // 查询文档 // EN: Query documents
            var queryOpts = QueryOptions()
            if let sortVal = options["sort"] as? [String: Any] {
                queryOpts.sort = convertToBSONDocument(sortVal)
            }
            if let limit = options["limit"] as? Int {
                queryOpts.limit = Int64(limit)
            }
            if let skip = options["skip"] as? Int {
                queryOpts.skip = Int64(skip)
            }
            if let projVal = options["projection"] as? [String: Any] {
                queryOpts.projection = convertToBSONDocument(projVal)
            }

            let docs: [BSONDocument]
            if !queryOpts.sort.isEmpty || queryOpts.limit > 0 || queryOpts.skip > 0 || !queryOpts.projection.isEmpty {
                docs = try await col.findWithOptions(filter, options: queryOpts)
            } else {
                docs = try await col.find(filter)
            }
            result.count = Int64(docs.count)

        case "findOne":
            // 查询单个文档 // EN: Query single document
            let doc = try await col.findOne(filter)
            result.count = doc != nil ? 1 : 0

        case "updateOne":
            // 更新单个文档 // EN: Update single document
            let upsert = options["upsert"] as? Bool ?? false
            let res = try await col.updateOne(filter, update: update, upsert: upsert)
            result.matched_count = res.matchedCount
            result.modified_count = res.modifiedCount

        case "updateMany":
            // 批量更新文档 // EN: Batch update documents
            let res = try await col.update(filter, update: update, upsert: false)
            result.matched_count = res.matchedCount
            result.modified_count = res.modifiedCount

        case "deleteOne":
            // 删除单个文档 // EN: Delete single document
            let count = try await col.deleteOne(filter)
            result.deleted_count = count

        case "deleteMany":
            // 批量删除文档 // EN: Batch delete documents
            let count = try await col.delete(filter)
            result.deleted_count = count

        case "replaceOne":
            // 替换单个文档 // EN: Replace single document
            let count = try await col.replaceOne(filter, replacement: doc)
            result.matched_count = count > 0 ? 1 : 0
            result.modified_count = count

        case "findAndModify":
            // 查找并修改文档 // EN: Find and modify document
            var cmd = BSONDocument()
            cmd["findAndModify"] = .string(tc.collection)
            cmd["query"] = .document(filter)
            if !update.isEmpty {
                cmd["update"] = .document(update)
            }
            if let newVal = options["new"] as? Bool {
                cmd["new"] = .bool(newVal)
            }
            if let upsert = options["upsert"] as? Bool {
                cmd["upsert"] = .bool(upsert)
            }
            if let remove = options["remove"] as? Bool {
                cmd["remove"] = .bool(remove)
            }
            let res = try await db.runCommand(cmd)
            if res["value"] != nil && res["value"] != .null {
                result.count = 1
            } else {
                result.count = 0
            }

        case "distinct":
            // 获取字段的不同值 // EN: Get distinct values of a field
            let field = options["field"] as? String ?? ""
            let values = try await col.distinct(field, filter: filter)
            result.count = Int64(values.count)

        case "aggregate":
            // 聚合管道操作 // EN: Aggregation pipeline operation
            if let pipeline = options["pipeline"] as? [[String: Any]] {
                let stages = pipeline.map { convertToBSONDocument($0) }
                let pipelineObj = try Pipeline(stages: stages, db: db)
                let inputDocs = try await col.find(BSONDocument())
                let outDocs = try await pipelineObj.execute(inputDocs)
                result.count = Int64(outDocs.count)
            }

        case "createIndex":
            // 创建索引 // EN: Create index
            if let keys = options["keys"] as? [String: Any] {
                let keysDoc = convertToBSONDocument(keys)
                var optsDoc = BSONDocument()
                if let indexOpts = options["options"] as? [String: Any] {
                    if let name = indexOpts["name"] as? String {
                        optsDoc["name"] = .string(name)
                    }
                    if let unique = indexOpts["unique"] as? Bool {
                        optsDoc["unique"] = .bool(unique)
                    }
                }
                _ = try await col.createIndex(keys: keysDoc, options: optsDoc)
                result.count = 1
            }

        case "listIndexes":
            // 列出索引 // EN: List indexes
            let indexes = await col.listIndexes()
            result.count = Int64(indexes.count)

        case "dropIndex":
            // 删除索引 // EN: Drop index
            if let name = options["name"] as? String {
                try await col.dropIndex(name)
            }

        default:
            throw RunnerError.unknownMethod(action.method)
        }
    }

    // MARK: - BSON 转换 / BSON Conversion

    /// 将字典转换为 BSON 文档
    /// EN: Convert dictionary to BSON document
    /// - Parameter value: 要转换的值 / Value to convert
    /// - Returns: BSON 文档 / BSON document
    private func convertToBSONDocument(_ value: Any) -> BSONDocument {
        guard let dict = value as? [String: Any] else {
            return BSONDocument()
        }

        var doc = BSONDocument()
        for (key, val) in dict {
            doc[key] = convertToBSONValue(val)
        }
        return doc
    }

    /// 将任意值转换为 BSON 值
    /// EN: Convert any value to BSON value
    /// - Parameter value: 要转换的值 / Value to convert
    /// - Returns: BSON 值 / BSON value
    private func convertToBSONValue(_ value: Any) -> BSONValue {
        switch value {
        case is NSNull:
            return .null
        case let bool as Bool:
            return .bool(bool)
        case let int as Int:
            return .int64(Int64(int))
        case let int64 as Int64:
            return .int64(int64)
        case let double as Double:
            return .double(double)
        case let string as String:
            return .string(string)
        case let array as [Any]:
            let bsonArray = BSONArray(array.map { convertToBSONValue($0) })
            return .array(bsonArray)
        case let dict as [String: Any]:
            return .document(convertToBSONDocument(dict))
        default:
            return .null
        }
    }
}

// MARK: - 错误类型 / Error Types

/// 运行器错误枚举
/// EN: Runner error enumeration
enum RunnerError: Error, CustomStringConvertible {
    /// 数据库未打开
    /// EN: Database not opened
    case databaseNotOpened
    /// 未知方法
    /// EN: Unknown method
    case unknownMethod(String)
    /// 连接失败
    /// EN: Connection failed
    case connectionFailed(String)

    /// 错误描述
    /// EN: Error description
    var description: String {
        switch self {
        case .databaseNotOpened:
            return "Database not opened"
        case .unknownMethod(let method):
            return "Unknown method: \(method)"
        case .connectionFailed(let msg):
            return "Connection failed: \(msg)"
        }
    }
}

File: runner/dart/lib/api_runner.dart
// Created by Yanjunhui

import 'package:monolite/monolite.dart';
import 'types.dart';

/// API 模式测试运行器
/// EN: APIRunner tests using the library API directly.
class APIRunner {
  final Database db; // 数据库实例

  APIRunner(this.db);

  /// 创建 API 运行器
  /// EN: Creates an API runner.
  static Future<APIRunner> create(String dbPath) async {
    final db = await Database.open(dbPath);
    return APIRunner(db);
  }

  /// 关闭数据库
  /// EN: Closes the database connection.
  Future<void> close() async {
    await db.close();
  }

  /// 运行单个测试
  /// EN: Runs a single test case.
  Future<TestResult> runTest(TestCase tc) async {
    final stopwatch = Stopwatch()..start();
    final result = TestResult(
      testName: tc.name,
      language: 'dart',
      mode: 'api',
    );

    try {
      // 执行前置步骤
      await _executeSetup(tc);

      // 执行测试动作
      await _executeAction(tc, result);

      result.success = true;
    } catch (e) {
      result.error = e.toString();
      // 检查是否是预期的错误
      if (tc.expected.error != null && tc.expected.error!.isNotEmpty) {
        result.success = true;
      }
    }

    stopwatch.stop();
    result.durationMs = stopwatch.elapsedMilliseconds;
    return result;
  }

  /// 执行前置步骤
  /// EN: Executes setup steps.
  Future<void> _executeSetup(TestCase tc) async {
    if (tc.setup.isEmpty) return;

    final col = await db.collection(tc.collection);

    for (final step in tc.setup) {
      switch (step.operation) {
        case 'insert':
          final doc = toBsonDocument(step.data);
          await col.insert([doc]);
          break;
        case 'createIndex':
          final opts = step.data as Map<String, dynamic>;
          final keys = opts['keys'] as Map<String, dynamic>;
          final indexOpts = opts['options'] as Map<String, dynamic>?;
          final unique = indexOpts?['unique'] as bool? ?? false;
          await col.createIndex(
            keys.map((k, v) => MapEntry(k, (v as num).toInt())),
            unique: unique,
          );
          break;
      }
    }
  }

  /// 执行测试动作
  /// EN: Executes test action.
  Future<void> _executeAction(TestCase tc, TestResult result) async {
    final col = await db.collection(tc.collection);

    switch (tc.action.method) {
      case 'insertOne':
        await _executeInsertOne(col, tc, result);
        break;
      case 'insertMany':
        await _executeInsertMany(col, tc, result);
        break;
      case 'find':
        await _executeFind(col, tc, result);
        break;
      case 'findOne':
        await _executeFindOne(col, tc, result);
        break;
      case 'updateOne':
        await _executeUpdateOne(col, tc, result);
        break;
      case 'updateMany':
        await _executeUpdateMany(col, tc, result);
        break;
      case 'deleteOne':
        await _executeDeleteOne(col, tc, result);
        break;
      case 'deleteMany':
        await _executeDeleteMany(col, tc, result);
        break;
      case 'replaceOne':
        await _executeReplaceOne(col, tc, result);
        break;
      case 'count':
        _executeCount(col, tc, result);
        break;
      case 'createIndex':
        await _executeCreateIndex(col, tc, result);
        break;
      case 'listIndexes':
        _executeListIndexes(col, tc, result);
        break;
      case 'dropIndex':
        await _executeDropIndex(col, tc, result);
        break;
      case 'findAndModify':
        await _executeFindAndModify(col, tc, result);
        break;
      case 'distinct':
        await _executeDistinct(col, tc, result);
        break;
      case 'aggregate':
        await _executeAggregate(col, tc, result);
        break;
      default:
        throw UnsupportedError('未知方法: ${tc.action.method}');
    }
  }

  /// 执行插入单个文档
  Future<void> _executeInsertOne(Collection col, TestCase tc, TestResult result) async {
    final doc = toBsonDocument(tc.action.doc);
    await col.insert([doc]);
    result.count = 1;
  }

  /// 执行批量插入文档
  Future<void> _executeInsertMany(Collection col, TestCase tc, TestResult result) async {
    final docs = <BsonDocument>[];
    for (final item in tc.action.docs ?? []) {
      docs.add(toBsonDocument(item));
    }
    await col.insert(docs);
    result.count = docs.length;
  }

  /// 执行查询
  Future<void> _executeFind(Collection col, TestCase tc, TestResult result) async {
    BsonDocument? filter;
    if (tc.action.filter != null) {
      filter = toBsonDocument(tc.action.filter);
    }

    QueryOptions? opts;
    if (tc.action.options != null) {
      final optMap = tc.action.options as Map<String, dynamic>;
      opts = QueryOptions(
        sort: toSortMap(optMap['sort']),
        limit: optMap['limit'] as int?,
        skip: optMap['skip'] as int?,
        projection: toSortMap(optMap['projection']),
      );
    }

    List<BsonDocument> docs;
    if (opts != null) {
      docs = await col.findWithOptions(filter, opts);
    } else {
      docs = await col.find(filter);
    }

    result.count = docs.length;
    result.documents = docs.map((d) => documentToMap(d)).toList();
  }

  /// 执行查询单个文档
  Future<void> _executeFindOne(Collection col, TestCase tc, TestResult result) async {
    final filter = toBsonDocument(tc.action.filter);
    final doc = await col.findOne(filter);
    if (doc != null) {
      result.count = 1;
      result.documents = [documentToMap(doc)];
    } else {
      result.count = 0;
    }
  }

  /// 执行更新单个文档
  Future<void> _executeUpdateOne(Collection col, TestCase tc, TestResult result) async {
    final filter = toBsonDocument(tc.action.filter);
    final update = toBsonDocument(tc.action.update);

    final updateResult = await col.update(filter, update);
    result.matchedCount = updateResult.matchedCount;
    result.modifiedCount = updateResult.modifiedCount;
    result.upsertedId = updateResult.upsertedId;
  }

  /// 执行更新多个文档
  Future<void> _executeUpdateMany(Collection col, TestCase tc, TestResult result) async {
    final filter = toBsonDocument(tc.action.filter);
    final update = toBsonDocument(tc.action.update);

    final updateResult = await col.update(filter, update);
    result.matchedCount = updateResult.matchedCount;
    result.modifiedCount = updateResult.modifiedCount;
  }

  /// 执行删除单个文档
  Future<void> _executeDeleteOne(Collection col, TestCase tc, TestResult result) async {
    final filter = toBsonDocument(tc.action.filter);
    final count = await col.deleteOne(filter);
    result.deletedCount = count;
  }

  /// 执行删除多个文档
  Future<void> _executeDeleteMany(Collection col, TestCase tc, TestResult result) async {
    final filter = toBsonDocument(tc.action.filter);
    // 使用 delete 方法代替 deleteMany
    final count = await col.delete(filter);
    result.deletedCount = count;
  }

  /// 执行替换单个文档
  Future<void> _executeReplaceOne(Collection col, TestCase tc, TestResult result) async {
    final filter = toBsonDocument(tc.action.filter);
    final replacement = toBsonDocument(tc.action.doc);

    // 先找到文档
    final existing = await col.findOne(filter);
    if (existing != null) {
      // 保留 _id
      final id = existing['_id'];
      if (id != null) {
        replacement['_id'] = id;
      }
      // 删除旧文档并插入新文档
      await col.deleteOne(filter);
      await col.insert([replacement]);
      result.matchedCount = 1;
      result.modifiedCount = 1;
    } else {
      result.matchedCount = 0;
      result.modifiedCount = 0;
    }
  }

  /// 执行计数
  void _executeCount(Collection col, TestCase tc, TestResult result) {
    result.count = col.count();
  }

  /// 执行创建索引
  Future<void> _executeCreateIndex(Collection col, TestCase tc, TestResult result) async {
    final opts = tc.action.options as Map<String, dynamic>;
    final keys = opts['keys'] as Map<String, dynamic>;
    final indexOpts = opts['options'] as Map<String, dynamic>?;
    final unique = indexOpts?['unique'] as bool? ?? false;
    final name = indexOpts?['name'] as String?;

    final indexName = await col.createIndex(
      keys.map((k, v) => MapEntry(k, (v as num).toInt())),
      name: name,
      unique: unique,
    );
    result.count = 1;
    result.documents = [{'indexName': indexName}];
  }

  /// 执行列出索引
  void _executeListIndexes(Collection col, TestCase tc, TestResult result) {
    final indexes = col.listIndexes();
    result.count = indexes.length;
  }

  /// 执行删除索引
  Future<void> _executeDropIndex(Collection col, TestCase tc, TestResult result) async {
    final opts = tc.action.options as Map<String, dynamic>;
    final name = opts['name'] as String;
    await col.dropIndex(name);
  }

  /// 执行 findAndModify
  Future<void> _executeFindAndModify(Collection col, TestCase tc, TestResult result) async {
    final filter = tc.action.filter != null ? toBsonDocument(tc.action.filter) : null;
    final optMap = tc.action.options as Map<String, dynamic>?;

    final isRemove = optMap?['remove'] as bool? ?? false;
    final returnNew = optMap?['new'] as bool? ?? false;

    if (isRemove) {
      // 删除模式
      final doc = await col.findOne(filter);
      if (doc != null) {
        await col.deleteOne(filter);
        result.count = 1;
        result.documents = [documentToMap(doc)];
      } else {
        result.count = 0;
      }
    } else {
      // 更新模式
      final update = tc.action.update != null ? toBsonDocument(tc.action.update) : null;

      if (returnNew) {
        // 先更新，再返回新文档
        if (update != null) {
          await col.update(filter, update);
        }
        final doc = await col.findOne(filter);
        if (doc != null) {
          result.count = 1;
          result.documents = [documentToMap(doc)];
        } else {
          result.count = 0;
        }
      } else {
        // 先返回旧文档，再更新
        final doc = await col.findOne(filter);
        if (doc != null) {
          result.count = 1;
          result.documents = [documentToMap(doc)];
          if (update != null) {
            await col.update(filter, update);
          }
        } else {
          result.count = 0;
        }
      }
    }
  }

  /// 执行 distinct
  Future<void> _executeDistinct(Collection col, TestCase tc, TestResult result) async {
    final filter = tc.action.filter != null ? toBsonDocument(tc.action.filter) : null;
    final optMap = tc.action.options as Map<String, dynamic>?;
    final field = optMap?['field'] as String? ?? '';

    final values = await col.distinct(field, filter);
    result.count = values.length;
  }

  /// 执行聚合管道（简化实现）
  Future<void> _executeAggregate(Collection col, TestCase tc, TestResult result) async {
    final optMap = tc.action.options as Map<String, dynamic>?;
    final pipeline = optMap?['pipeline'] as List<dynamic>? ?? [];

    // 从集合获取所有文档作为起点
    var docs = await col.find(null);

    // 简化的聚合管道处理
    for (final stage in pipeline) {
      final stageDoc = toBsonDocument(stage);

      if (stageDoc.containsKey('\$match')) {
        // $match 阶段
        final matchFilter = toBsonDocument(stageDoc['\$match']);
        docs = docs.where((doc) => matchesFilter(doc, matchFilter)).toList();
      } else if (stageDoc.containsKey('\$sort')) {
        // $sort 阶段
        final sortSpec = toSortMap(stageDoc['\$sort']);
        if (sortSpec != null) {
          docs = _sortDocuments(docs, sortSpec);
        }
      } else if (stageDoc.containsKey('\$limit')) {
        // $limit 阶段
        final limit = stageDoc['\$limit'] as int;
        if (docs.length > limit) {
          docs = docs.sublist(0, limit);
        }
      } else if (stageDoc.containsKey('\$skip')) {
        // $skip 阶段
        final skip = stageDoc['\$skip'] as int;
        if (docs.length > skip) {
          docs = docs.sublist(skip);
        } else {
          docs = [];
        }
      } else if (stageDoc.containsKey('\$project')) {
        // $project 阶段
        final projSpec = toSortMap(stageDoc['\$project']);
        if (projSpec != null) {
          docs = _applyProjection(docs, projSpec);
        }
      } else if (stageDoc.containsKey('\$group')) {
        // $group 阶段（简化实现）
        final groupSpec = toBsonDocument(stageDoc['\$group']);
        docs = _executeGroup(docs, groupSpec);
      }
    }

    result.count = docs.length;
    result.documents = docs.map((d) => documentToMap(d)).toList();
  }

  /// 对文档排序
  List<BsonDocument> _sortDocuments(List<BsonDocument> docs, Map<String, int> sortSpec) {
    final result = List<BsonDocument>.from(docs);
    result.sort((a, b) {
      for (final entry in sortSpec.entries) {
        final valA = a[entry.key];
        final valB = b[entry.key];
        final cmp = _compareValues(valA, valB);
        if (cmp != 0) {
          return entry.value < 0 ? -cmp : cmp;
        }
      }
      return 0;
    });
    return result;
  }

  /// 比较两个值
  int _compareValues(dynamic a, dynamic b) {
    if (a == null && b == null) return 0;
    if (a == null) return -1;
    if (b == null) return 1;
    if (a is num && b is num) return a.compareTo(b);
    if (a is String && b is String) return a.compareTo(b);
    return a.toString().compareTo(b.toString());
  }

  /// 应用投影
  List<BsonDocument> _applyProjection(List<BsonDocument> docs, Map<String, int> projection) {
    final result = <BsonDocument>[];

    // 判断是包含还是排除模式
    var includeMode = false;
    for (final entry in projection.entries) {
      if (entry.key == '_id') continue;
      if (entry.value == 1) {
        includeMode = true;
      }
      break;
    }

    for (final doc in docs) {
      if (includeMode) {
        final newDoc = BsonDocument();
        // 默认包含 _id
        var includeId = true;
        if (projection['_id'] == 0) includeId = false;
        if (includeId && doc.containsKey('_id')) {
          newDoc['_id'] = doc['_id'];
        }
        for (final entry in projection.entries) {
          if (entry.key == '_id') continue;
          if (entry.value == 1 && doc.containsKey(entry.key)) {
            newDoc[entry.key] = doc[entry.key];
          }
        }
        result.add(newDoc);
      } else {
        final newDoc = BsonDocument();
        final excludeFields = <String>{};
        for (final entry in projection.entries) {
          if (entry.value == 0) excludeFields.add(entry.key);
        }
        for (final entry in doc.entries) {
          if (!excludeFields.contains(entry.key)) {
            newDoc[entry.key] = entry.value;
          }
        }
        result.add(newDoc);
      }
    }
    return result;
  }

  /// 执行 $group 阶段
  List<BsonDocument> _executeGroup(List<BsonDocument> docs, BsonDocument groupSpec) {
    final idSpec = groupSpec['_id'];
    final groups = <String, List<BsonDocument>>{};

    // 按 _id 分组
    for (final doc in docs) {
      String groupKey;
      if (idSpec == null) {
        groupKey = 'null';
      } else if (idSpec is String && idSpec.startsWith('\$')) {
        final field = idSpec.substring(1);
        groupKey = doc[field]?.toString() ?? 'null';
      } else {
        groupKey = idSpec.toString();
      }

      groups.putIfAbsent(groupKey, () => []).add(doc);
    }

    // 构建结果
    final result = <BsonDocument>[];
    for (final entry in groups.entries) {
      final groupDoc = BsonDocument();
      groupDoc['_id'] = entry.key == 'null' ? null : entry.key;

      // 处理聚合操作符
      for (final specEntry in groupSpec.entries) {
        if (specEntry.key == '_id') continue;

        final accSpec = specEntry.value;
        if (accSpec is BsonDocument) {
          if (accSpec.containsKey('\$sum')) {
            final sumSpec = accSpec['\$sum'];
            if (sumSpec == 1) {
              groupDoc[specEntry.key] = entry.value.length;
            } else if (sumSpec is String && sumSpec.startsWith('\$')) {
              final field = sumSpec.substring(1);
              num sum = 0;
              for (final d in entry.value) {
                final val = d[field];
                if (val is num) sum += val;
              }
              groupDoc[specEntry.key] = sum;
            }
          } else if (accSpec.containsKey('\$avg')) {
            final avgSpec = accSpec['\$avg'];
            if (avgSpec is String && avgSpec.startsWith('\$')) {
              final field = avgSpec.substring(1);
              num sum = 0;
              int count = 0;
              for (final d in entry.value) {
                final val = d[field];
                if (val is num) {
                  sum += val;
                  count++;
                }
              }
              groupDoc[specEntry.key] = count > 0 ? sum / count : 0;
            }
          }
        }
      }

      result.add(groupDoc);
    }

    return result;
  }
}
